lib/Utils/Paths.cpp

```diff
@@ -0,0 +1,229 @@
+
+        if (!m_WithSystem) {
+          if (E.Group != frontend::After || E.Group != frontend::Quoted)
+            continue;
+        }
+        if (m_WithFlags) {
+          switch (E.Group) {
+            case frontend::After:
+              addFlag("-idirafter");
+              break;
+
+            case frontend::Quoted:
+              addFlag("-iquote");
+              break;
+
+            case frontend::System:
+              addFlag("-isystem");
+              break;
+
+            case frontend::IndexHeaderMap:
+              addFlag("-index-header-map");
+              addFlag(E);
+              break;
+
+            case frontend::CSystem:
+              addFlag("-c-isystem");
+              break;
+
+            case frontend::ExternCSystem:
+              addFlag("-extern-c-isystem");
+              break;
+
+            case frontend::CXXSystem:
+              addFlag("-cxx-isystem");
+              break;
+
+            case frontend::ObjCSystem:
+              addFlag("-objc-isystem");
+              break;
+
+            case frontend::ObjCXXSystem:
+              addFlag("-objcxx-isystem");
+              break;
+
+            case frontend::Angled:
+              addFlag(E);
+              break;
+          }
+        }
+
+        addPath(E.Path);
+      }
+
+      if (m_WithSystem) {
+        if (!m_Opts.ResourceDir.empty()) {
+          addArg("-resource-dir");
+          addPath(m_Opts.ResourceDir);
+        }
+        if (!m_Opts.ModuleCachePath.empty()) {
+          addArg("-fmodule-cache-path");
+          addPath(m_Opts.ModuleCachePath);
+        }
+        if (!m_Opts.UseStandardSystemIncludes)
+          addArg("-nostdinc");
+        if (!m_Opts.UseStandardCXXIncludes)
+          addArg("-nostdinc++");
+        if (m_Opts.UseLibcxx)
+          addArg("-stdlib=libc++");
+        if (m_Opts.Verbose)
+          addArg("-v");
+      }
+    }
+  };
+
+  class StreamDump : public FormatIncludes {
+    llvm::raw_ostream& m_Stream;
+  public:
+    StreamDump(const HeaderSearchOptions& Opts, bool WithSystem, bool WithFlags,
+               llvm::raw_ostream& S) :
+      FormatIncludes(Opts, WithSystem, WithFlags), m_Stream(S) {}
+
+    ~StreamDump() { m_Stream << "\n"; }
+    void addFlag(const char* Arg) override { m_Stream << Arg << " "; }
+    void addPath(const std::string& Str) override {  m_Stream << Str << "\n"; }
+  };
+
+  class VectorCopy : public FormatIncludes {
+    SmallVectorImpl<std::string>& m_Vector;
+  public:
+    VectorCopy(const HeaderSearchOptions& Opts, bool WithSystem, bool WithFlags,
+               SmallVectorImpl<std::string>& V) :
+      FormatIncludes(Opts, WithSystem, WithFlags), m_Vector(V) {}
+
+    void addFlag(const char* Arg) override { m_Vector.push_back(Arg); }
+    void addPath(const std::string& Str) override {  m_Vector.push_back(Str); }
+  };
+}
+
+namespace cling {
+namespace utils {
+
+void CopyIncludePaths(const clang::HeaderSearchOptions& Opts,
```

Comment:
I think this can be implemented by adding the 2 functions in the base class and remove the derived classes. Makes it more maintainable and consistent to the rest of the code base. I don't have issues with java-style decorators/hierarchies but we are trying to avoid them as much as possible.